### DISCUSSION TITLE
How do you see payment service integration? - e.g. PayPal

### DISCUSSION THREAD

**Discussion by func0der** [OpenMage/magento-lts#2322] [Questions & Answers]
Hey,

I was wondering if you guys have any plan/idea on the future of payment service integrations.

A little bit of history to this question.

PayPal is currently pushing their new [PayPal Checkout integration](https://developer.paypal.com/docs/checkout/). 
The has been no official integration for Magento 2, at least as far a the narrative of PayPal goes (https://www.paypal.com/en/webapps/mpp/upgrade-to-checkout#showcase1).

Magento 1 was using or better yet is using PayPal's NVP API in its codebase, which has been deprecated for years AFAIK.
I-Ways GmbH had provided a Magento 1 extensions for PayPal Plus for the last couple of years. Someone has provided it as a [composer package](https://packagist.org/packages/kgoedecke/magento-paypal-plus-extension). But since PayPal wants you to stop using that, too, it soon will either stop working or stop supporting certain payment methods.

Since OpenMage is not as widely adopted as Magento 1 or even has a platform big enough to be considered by PayPal for a listing on the above mentioned page yet, we are having a bit of an issue here.

This is of course not only related to PayPal, but to other payment solutions as well. I chose PayPal here, as their are notorious for their every-other-year-rewrite-of-their-whole-api. :D 

There are some question I want to through out there:

1. What is OpenMage?
    a. Is a simple LTS version of Magento 1, which mitigates the 'immediate' thread of high costs to migrate to another platform?
    b. A project that plans to become a standalone shop system?

2. If 1b: How does OpenMage want to implement payment service in the future?
    a. Should 'essentials' like PayPal be part of the core and 'exotic' ones extensions?
    b. or should all external payment services be extensions?

3. If 1b: How would OpenMage deal with the very deprecated and probably dangerous to use default implementations of payment providers, that are currently part of the core?

Thanks for your time.

**Comment by justinbeaty** [comment]
I only speak for myself, but here's some my opinion:

1. Right now it's both I guess. 1.9.4.x branch for immediate migration and 20.0 for something that will continue to grow.
2. I think all payment services could be stripped. It is less for us to maintain.

At some point I think we need more compatible modules listed on openmage.org. For example, I have a Gene/Braintree fork[1] that uses composer to install the latest major version of the `braintree_php` SDK. It looks like the module is still loading an older version of the JS SDK (3.48.0) vs the newest (3.85.3) but I am actually using Magento headless, so I haven't updated that. But it's not a major version difference so I don't think Braintree is removing that option any time soon.

In any case, I just think it's probably easiest to have payment modules broken out, and have them listed somewhere that people can find them.

[1] https://github.com/justinbeaty/module-gene-braintree

**Comment by Flyingmana** [comment]
We are likely to become standalone in the long run.

Only speaking about my personal opinion on this.
OpenMage should definitive have a way to support up2date PayPal and other commonly used payment platforms.
It doesnt neccesarily mean, it has to be part of the out of the box installation.

For a healthy ecosystem its useful for other developers and companies to provide their own modules without having to compete with OpenMage directly, which at the same times means also less work for our core Team.


That said, if the ecosystem is not going to provide a working paypal integration, then we should provide our own,

**Comment by JonLaliberte** [comment]
As far as I can tell, most of the large module makers, as well as payment services (PayPal, Stripe) are ignoring LTS and focusing on M2. That's too bad, but also the current state of things. Without these folks it's a bit more difficult to keep these types of modules up to date and well maintained.

It seems to me that PayPal, and maybe a maintained Stripe (and/or Auth.net compatible option) would be the minimum that I'd like to see OpenMage handling. If not included in the base install, they could be maintained by OpenMage and distributed as modules you can install later.

I'm not sure the best thing though either, if they aren't going to be maintained (especially if they have known security issues), they should probably not be in LTS.

**Comment by kiatng** [comment]
I maintain [ipay88](https://www.ipay88.com/) credit card for many years now.

**Comment by func0der** [comment]
I am not seeing PayPal Checkout being included any time soon.
It basically jumps out of your checkout process instantaneously if you chose one of their payment methods. Without making payment the last step in the OnePage checkout, this would only work with a pretty hacky workaround. I have yet to find a shop that is build like OpenMage in the checkout process and has a working PayPal Checkout integration.

You could "reserve" funds for your shop, but even if PayPal claims that it will refund the funds after a reasonable short period of time, I would not burden small to medium sized shops with that weird behavior.

**Comment by JonLaliberte** [comment]
Isn't that the same as what PayPal Express does? Never had a problem with that one.

**Reply by func0der** [reply]
PayPal Express is the checkout directly on the product page, isn't it?

**Reply by JonLaliberte** [reply]
Express is the one on the Cart page that brings you to PayPal to checkout. After logging in to PayPal you get sent back to Magento to confirm the order.

**Reply by func0der** [reply]
Okay, I you jump from product or cart page should not make a big difference.
But if you are coming from the one page checkout, you can not just leave the checkout process at let the customer 'pay' without showing them a summary of what it is they are buying. This is a least the case in some countries.
That's why I am concerned.

**Comment by alexh-swdev** [comment]
I was wondering how/when OM will handle "PP Checkout"?

And @func0der: Did "Login and Pay with Amazon" not work the same/similar way, via a button in the cart and its own (Mage-) checkout-page?

**Reply by func0der** [reply]
I am not familiar with "Amazon" in ANY way :D Sorry, I can not comment on this. Sorry
